aten/src/ATen/native/RangeFactories.cpp

```diff
@@ -76,19 +77,33 @@ Tensor& logspace_cpu_out(Tensor& result, Scalar start, Scalar end, int64_t steps
 }
 
 Tensor& range_cpu_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
-  AT_DISPATCH_ALL_TYPES(result.scalar_type(), "range_cpu", [&]() {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(result.scalar_type(), "range_cpu", [&]() {
     using accscalar_t = at::acc_type<scalar_t, false>;
     auto xstart = start.to<accscalar_t>();
     auto xend = end.to<accscalar_t>();
     auto xstep = step.to<accscalar_t>();
 
-    TORCH_CHECK(xstep > 0 || xstep < 0, "step must be nonzero");
-    TORCH_CHECK(std::isfinite(static_cast<double>(xstart)) &&
-             std::isfinite(static_cast<double>(xend)),
-             "unsupported range: ", xstart, " -> ", xend);
-    TORCH_CHECK(((xstep > 0) && (xend >= xstart)) || ((xstep < 0) && (xend <= xstart)),
-             "upper bound and larger bound inconsistent with step sign");
-    int64_t size = static_cast<int64_t>(((xend - xstart) / xstep) + 1);
+    int64_t size;
+    if (std::is_same<scalar_t, std::complex<double>>::value || std::is_same<scalar_t, std::complex<float>>::value) {
+        TORCH_CHECK(std::abs(xstep) > 0 || std::abs(xstep) < 0, "step must be nonzero");
+        TORCH_CHECK(std::isfinite(static_cast<double>(std::abs(xstart))) &&
+                 std::isfinite(static_cast<double>(std::abs(xend))),
+                 "unsupported range: ", xstart, " -> ", xend);
+        TORCH_CHECK(((std::abs(xstep) > 0) && (std::abs(xend) >= std::abs(xstart))) ||
+                    ((std::abs(xstep) < 0) && (std::abs(xend) <= std::abs(xstart))),
+                 "upper bound and larger bound inconsistent with step sign");
+        size = static_cast<int64_t>(std::abs(((xend - xstart) / xstep)) + 1);
+    }
+    else{
+        TORCH_CHECK(std::real(xstep) > 0 || std::real(xstep) < 0, "step must be nonzero");
```

Comment:
Is using `real` here really correct? This means that negative steps don't work with complex numbers. Maybe you should be allowed to step in a direction that is not the real direction. Can we perhaps defer `range` for another PR? It is not altogether clear what its complex semantics should be